cs/src/io/unsafe/InputStream.cs

```diff
@@ -93,30 +96,45 @@ public override ArraySegment<byte> ReadBytes(int count)
 
         internal override void EndOfStream(int count)
         {
-            var oldBuffer = buffer;
-
+            // The unread bytes left in the buffer. May be negative, which
+            // indicates that this stream has been advanced beyond where we
+            // are in the underlying stream and some bytes will need to be
+            // skipped.
             var remaining = end - position;
 
-            if (remaining < 0)
-            {
-                stream.Seek(-remaining, SeekOrigin.Current);
-                remaining = 0;
-            }
-
             bool failed = false;
-            byte[][] tempBuffers = null;
-            int numTempBuffers = 0;
+            byte[][] tempBuffers = EmptyTempBuffers;
+
+            // Check whether we need to read in chunks to avoid allocating a
+            // ton of memory ahead of time.
             if ((count > buffer.Length && (count - buffer.Length > ActiveAllocationChunk)))
             {
-                // Calculate number of temp buffers -1 in case difference is exactly
-                // multiple of chunk size.
-                numTempBuffers = (count - buffer.Length - 1) / ActiveAllocationChunk;
+                // Calculate number of temp buffers - 1 in case difference is
+                // exactly multiple of chunk size. If it's exactly a
+                // multiple, we need one fewer temp buffers, as we'll read
+                // the last chunk directly into the final buffer, below.
```

Comment:
Comment is not really accurate (partly due to confused formatting on previous version). How about something like?
"Calculate number of temp buffers; we round down since last chunk is read directly into final buffer. Note: Difference is adjusted by -1 to round down correctly in case where difference is exactly multiple of allocation chunk size."